update_prevalence.py

```diff
@@ -281,16 +281,53 @@ def recent_daily_cumulative_cases(self) -> List[int]:
     # cases. Nominally is values[-1] - values[0], but sometimes regions post
     # corrections which result in the number of cases decreasing.
     def cases_in_cum_cases(self, values) -> int:
-        min_index = values.index(min(values))
-        return max(values[min_index:]) - min(values)
+        # list of indices right before negative corrections. If values = [3,2,3,0,5],
+        # then negative_corrections == [0, 2]
+        negative_corrections = [i for i, val in enumerate(values[:-1]) if val > values[i+1]]
+        corrections_within_bounds = min(values) == values[0] and max(values) == values[-1]
+        if len(negative_corrections) == 0 or corrections_within_bounds:
+            return values[-1] - values[0]
+
+        # Always use values[-1] rather than max(values) because max(values) is
+        # either values[-1] or there's been a negative correction that should take precedence.
+        if min(values[:-1]) <= values[-1]:
+            possibly_suspect_correction = False
+            for local_max_idx in negative_corrections:
+                value_before_correction = values[local_max_idx]
+                value_correction = values[local_max_idx + 1]
+                value_after_correction = None if local_max_idx + 2 >= len(values) else values[local_max_idx + 2]
+
+                if value_after_correction is not None:
+                    correction_size = value_before_correction - value_correction # flipped
+                    overall_change = value_after_correction - value_before_correction
+                    # arbitrary heuristic to look for cases where the negative correction
+                    # itself might be the data that's wrong. Warn if:
+                    #   * removing the negative correction would create consistent data,
+                    #   * the correction is substantial (more than 5 people) and
+                    #   * the correction is proportionally much larger than the overall change
```

Comment:
This logic seems a bit odd to me; if the cumulative cases go 5, 10, 0, 13, correction_size is -10, overall change is 3, and this will warn. But I'd look at that as the 0 is just funky.

Meanwhile 5, 10, 0, 14 is fine? Why?

Reply:
yeah, it’s not great. But this is very, very noisy without some filtering, and if it’s not off by that much (for negotiable definitions of “that much”, I don’t think manually checking and fixing it will have great impact.

the specific cases I wanted to catch are “oops somebody made a typo and the number is off by thousands”.

![1D81184D-359A-4B9A-B519-1231F8F61BED](https://user-images.githubusercontent.com/4390573/127517889-730050f2-f9e4-4c8e-af04-089aace1acff.jpeg)